jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt

```diff
@@ -212,6 +223,123 @@ class AwsResourceCacheTest {
+        verify(mockResource, times(1)).fetch(projectRule.project, US_WEST_1, CRED2)
+        verify(mockResource, times(1)).fetch(projectRule.project, US_WEST_2, CRED2)
+    }
+
+    @Test
+    fun whenACredentialIdIsModifiedItsEntriesAreRemovedFromTheCache() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+        getAllRegionAndCredPermutations()
+
+        ApplicationManager.getApplication().messageBus.syncPublisher(CredentialManager.CREDENTIALS_CHANGED).providerModified(CRED1)
+
+        getAllRegionAndCredPermutations()
+
+        verify(mockResource, times(2)).fetch(projectRule.project, US_WEST_1, CRED1)
+        verify(mockResource, times(2)).fetch(projectRule.project, US_WEST_2, CRED1)
+        verify(mockResource, times(1)).fetch(projectRule.project, US_WEST_1, CRED2)
+        verify(mockResource, times(1)).fetch(projectRule.project, US_WEST_2, CRED2)
+    }
+
+    @Test
+    fun cacheExposesBlockingApi() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+        assertThat(sut.getResourceNow(mockResource)).isEqualTo("hello")
+    }
+
+    @Test
+    fun cacheExposesBlockingApiWithRegionAndCred() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+        assertThat(sut.getResourceNow(mockResource, US_WEST_1, CRED1)).isEqualTo("hello")
+        verify(mockResource).fetch(projectRule.project, US_WEST_1, CRED1)
+    }
+
+    @Test
+    fun cacheExposesBlockingApiWhereExecutionExceptionIsUnwrapped() {
+        whenever(mockResource.fetch(any(), any(), any())).thenThrow(RuntimeException("boom"))
+        assertThatThrownBy { sut.getResourceNow(mockResource, timeout = Duration.ofMillis(5)) }
+            .isInstanceOf(RuntimeException::class.java)
+            .withFailMessage("boom")
+    }
+
+    @Test
+    fun cacheExposesBlockingApiWithTimeout() {
+        whenever(mockResource.fetch(any(), any(), any())).thenAnswer {
+            Thread.sleep(50)
+            "hello"
+        }
+        assertThatThrownBy { sut.getResourceNow(mockResource, timeout = Duration.ofMillis(5)) }.isInstanceOf(TimeoutException::class.java)
+    }
+
+    @Test
+    fun canConditionallyFetchOnlyIfAvailableInCache() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+
+        assertThat(sut.getResourceIfPresent(mockResource, US_WEST_1, CRED1)).isNull()
+        sut.getResource(mockResource, US_WEST_1, CRED1).value
+        assertThat(sut.getResourceIfPresent(mockResource, US_WEST_1, CRED1)).isEqualTo("hello")
+    }
+
+    @Test
+    fun canConditionallyFetchOnlyIfAvailableInCacheAndRespectExpiry() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+
+        val now = Instant.now()
+        whenever(mockClock.instant()).thenReturn(now)
+        sut.getResource(mockResource, US_WEST_1, CRED1).value
+
+        whenever(mockClock.instant()).thenReturn(now.plus(DEFAULT_EXPIRY).plusMillis(1))
```

Comment:
Subtract 1 milli to grantee it will always be in the past

Reply:
this is current time - I want to make sure the expiry is always in the past hence current time is the expiry + 1 ms....

Reply:
Ah, k